File: models/conta.py
from models.cliente import Cliente
from utils.helper import formata_moeda

class Conta:

    codigo = 1000

    def __init__(self, cliente):
        self.__numero = Conta.codigo
        self.__cliente = cliente
        self.__saldo = 0.0
        self.__limite = 100.00
        self.__saldo_total = self._calcula_saldo_total
        Conta.codigo += 1

    def __str__(self):
        return f'Numero da conta: {self.numero}\nCliente: {self.cliente.nome}\n' \
               f'Saldo Total: {formata_moeda(self.saldo_total)}'

    @property
    def numero(self):
        return self.__numero

    @property
    def cliente(self):
        return self.__cliente

    @property
    def saldo(self):
        return self.__saldo

    @saldo.setter
    def saldo(self, valor):
        self.__saldo = valor

    @property
    def limite(self):
        return self.__limite

    @limite.setter
    def limite(self, valor):
        self.__limite = valor

    @property
    def saldo_total(self):
        return self.__saldo_total

    @saldo_total.setter
    def saldo_total(self, valor):
        self.__saldo_total = valor

    @property
    def _calcula_saldo_total(self):
        return self.saldo + self.limite

    def depositar(self, valor):
        if valor > 0:
            self.saldo = self.saldo + valor
            self.saldo_total = self._calcula_saldo_total
            print('Deposito efetuado com sucesso')
        else:
            print('Erro ao realizar o deposito tente novamente')

    def sacar(self, valor):
        if valor > 0 and self.saldo_total >= valor:
            if self.saldo >= valor:
                self.saldo = self.saldo - valor
                self.saldo_total = self._calcula_saldo_total
            else:
                restante = self.saldo - valor
                self.limite = self.limite + restante
                self.saldo = 0
                self.saldo_total = self._calcula_saldo_total
            print('Saque realizado com sucesso')
        else:
            print('Erro ao realizar o saque tente novamente')

    def transferir(self, destino, valor):
        if valor > 0 and self.saldo_total >= valor:
            if self.saldo >= valor:
                self.saldo = self.saldo - valor
                self.saldo_total = self._calcula_saldo_total
                destino.saldo = destino.saldo + valor
                destino.saldo_total = destino._calcula_saldo_total
                print('Transferencia realizada com sucesso')
            else:
                restante = self.saldo - valor
                self.limite = self.limite + restante
                self.saldo = 0
                self.saldo_total = self._calcula_saldo_total
        else:
            print('Erro ao realizar a transferencia tente novamente')

File: src/model/Conta.java
package model;

import helper.Utils;

/**
Metodo construtor da classe Conta
*/
public class Conta {
	private static int contador = 1001;
	private int numero;
	private Cliente cliente;
	private Double saldo = 0.0;
	private Double limite = 0.0;
	private Double saldo_total;
	
	
	public Conta(Cliente cliente) {
		this.numero = contador;
		this.cliente = cliente;
		contador++;
		this.atualiza_saldo_total();
	}
	
	/*
	Getters e Setters
	*/
	
	public Cliente getCliente() {
		return cliente;
	}
	
	public void setCliente(Cliente cliente) {
		this.cliente = cliente;
	}
	
	public Double getSaldo() {
		return saldo;
	}
	
	public void setSaldo(Double saldo) {
		this.saldo = saldo;
	}
	
	public Double getLimite() {
		return limite;
	}
	
	public void setLimite(Double limite) {
		this.limite = limite;
		this.atualiza_saldo_total();
	}
	
	public int getNumero() {
		return numero;
	}
	
	public Double getSaldo_total() {
		return saldo_total;
	}
	
	private void atualiza_saldo_total() {
		this.saldo_total = this.getSaldo() + this.getLimite();
	}
	
	
	/**
	 Sobrecarga do metodo "To String" da classe Conta
	 */
	public String toString() {
		return "Numero da conta: " + this.getNumero() + "\nCliente: " + this.cliente.getNome() +
				"\nSaldo Total: " + Utils.double_string(this.getSaldo_total());
	}
	
	/**
	 Metodo para depositar um valor dentro de uma conta
	 */
	public void depositar(Double valor) {
		if (valor > 0) {
			this.saldo = this.getSaldo() + valor;
			this.atualiza_saldo_total();
			System.out.println("Deposito efetuado com sucesso");
		}else {
			System.out.println("Erro ao efetuar deposito tente novamente");
		}
	}
	
	/**
	 Metodo para realizar um saque, caso seja possivel, subtrai o valor solicitado
	 */
	public void sacar(Double valor) {
		if (valor > 0 && this.getSaldo_total() >= valor) {
			if(this.getSaldo() > valor) {
				this.saldo = this.getSaldo() - valor;
				this.atualiza_saldo_total();
				System.out.println("Saque realizado com sucesso");
			}else {
				Double restante = this.getSaldo() - valor;
				this.limite = this.getLimite() + restante;
				this.saldo = 0.0;
				this.atualiza_saldo_total();
				System.out.println("Saque realizado com sucesso");
			}
		}else {
			System.out.println("Saque não realizado tente novamente");
		}
	}
	
	/**
	 Metodo para realizar uma transferencia, caso seja possivel, coloca o valor na conta destino
	 e subtrai da conta de origem
	 */
	public void transferir(Conta destino, Double valor) {
		if (valor > 0 && this.getSaldo_total() >= valor) {
			if(this.getSaldo() > valor) {
				this.saldo = this.getSaldo() - valor;
				destino.saldo = destino.getSaldo() + valor;
				this.atualiza_saldo_total();
				destino.atualiza_saldo_total();
				System.out.println("Transferencia realizada com sucesso");	
			}else {
				Double restante = this.getSaldo() - valor;
				this.limite = this.getLimite() + restante;
				this.saldo = 0.0;
				destino.saldo = destino.getSaldo() + valor;
				this.atualiza_saldo_total();
				destino.atualiza_saldo_total();
				System.out.println("Transferencia realizada com sucesso");
			}
	}else {
		System.out.println("Saque não realizado tente novamente");
	}
	}
}
		